src/librbd/io/ObjectRequest.cc

```diff
@@ -214,11 +215,17 @@ void ObjectReadRequest<I>::read_object() {
   ldout(image_ctx->cct, 20) << dendl;
 
   neorados::ReadOp read_op;
-  if (this->m_object_len >= image_ctx->sparse_read_threshold_bytes) {
-    read_op.sparse_read(this->m_object_off, this->m_object_len, m_read_data,
-                        m_extent_map);
-  } else {
-    read_op.read(this->m_object_off, this->m_object_len, m_read_data);
+  for (auto [object_off, object_len]: this->m_extents) {
+    auto read_data = std::make_unique<ceph::bufferlist>();
```

Comment:
Are these heap allocations necessary? If you first `m_extent_results.emplace_back({}, {})`, then `auto& extent_result = m_extent_results.back()`, you can then pass the pointers to the `extent_result.first` and `extent_result.second` to the `read_op`

Reply:
This works: `m_extent_results.emplace_back(std::make_unique<ceph::bufferlist>(), std::make_unique<Extents>());`,
but `m_extent_results.emplace_back({}, {});` does not compile.
Did you mean in addition to change the type of `m_extent_results` to something else?

Reply:
Yes, I was thinking your `m_extent_results` would be just a `std::vector<std::pair<bufferlist, Extents>>`. If the compiler cannot deduce the types, you could just use `m_extent_results.emplace_back(bufferlist{}, Extents{})`

Reply:
This was actually what I tried on the first place.
However, I got a segfault when when the mock read function tried to append to the buffer given to the read_op.
My guess is that the objects of the vector were re-alloced, and the pointer sent to the read_op was no longer valid.
This is why I moved to using a vector of pointers instead of vector of objects.
Maybe I can overcome this by reserving space on the vector before I actually emplace back objects.
What do you think?

Reply:
It would be good to `reserve` prior to the `emplace_back`, but it shouldn't be necessary. The only way that vector should have been invalidated would have been if the `ObjectReadRequest` instance was destructed prior to the mock being invoked. You could try running `CEPH_LIB=lib valgrind ./bin/unittest_librbd --gtest_filter=...` to see how the invalid access occurs.

Reply:
OK I just verified that indeed the vector is actually re-alloced when doing the second insert, thus invalidating the pointer to the first object.
Reserving on the vector prior to inserting solves this issue, and seem a sound solution I think as a vector should not be re-allocated if no more appends over its capacity occur.